src/platform/windows/display_vram.cpp

```diff
@@ -830,6 +755,202 @@ namespace platf::dxgi {
+          return -1;
+        }
+
+        // Get the texture from the mapped frame
+        frame_texture = (ID3D11Texture2D *) d3d11_frame->data[0];
+      }
+      else {
+        // Otherwise, we can just use the texture inside the original frame
+        frame_texture = (ID3D11Texture2D *) frame->data[0];
+      }
+
+      return base.init_output(frame_texture, frame->width, frame->height);
+    }
+
+  private:
+    d3d_base_encoding_stream_t base;
+    frame_t hwframe;
+  };
+
+  class d3d_nvenc_encoding_stream_t: public nvenc_encoding_stream_t {
+  public:
+    bool
+    init(std::shared_ptr<platf::display_t> display, adapter_t::pointer adapter_p, const ::video::config_t &client_config, pix_fmt_e pix_fmt) {
+      auto buffer_format = nvenc::nvenc_format_from_sunshine_format(pix_fmt);
+      if (buffer_format == NV_ENC_BUFFER_FORMAT_UNDEFINED) {
+        BOOST_LOG(error) << "Unexpected pixel format for NvENC ["sv << from_pix_fmt(pix_fmt) << ']';
+        return false;
+      }
+
+      if (base.init(display, adapter_p, pix_fmt)) return false;
+
+      // TODO: add sunshine colorspace/primaries/tranfer/matrix constants and translate those to avcodec, nvenc, whatever
+      {
+        int color_range = (client_config.encoderCscMode & 0x1) ? 2 : 1;
+        int color_space;
+        if (client_config.dynamicRange && display->is_hdr()) {
+          // When HDR is active, that overrides the colorspace the client requested
+          color_space = 9;
+        }
+        else {
+          switch (client_config.encoderCscMode >> 1) {
+            case 0:
+            default:
+              // BT.601
+              color_space = 5;
+              break;
+
+            case 1:
+              // BT.709
+              color_space = 1;
+              break;
+
+            case 2:
+              // BT.2020
+              color_space = 9;
+              break;
+          }
+        }
+        set_colorspace(color_space, color_range);
+      }
+
+      nvenc = std::make_unique<nvenc::nvenc_d3d11>(base.device.get(), client_config.width, client_config.height, buffer_format);
+
+      auto nvenc_colorspace = nvenc::nvenc_colorspace_from_rtsp_csc(client_config.encoderCscMode, client_config.dynamicRange && display->is_hdr());
+
+      nvenc::nvenc_config nvenc_config;
+      if (config::video.nv.nv_preset) nvenc_config.quality_preset = *config::video.nv.nv_preset - 11;
+
+      if (!nvenc->create_encoder(nvenc_config, client_config, nvenc_colorspace)) return false;
+
+      return base.init_output(nvenc->get_input_texture(), client_config.width, client_config.height) == 0;
+    }
+
+    encoded_frame
+    encode_frame(uint64_t frame_index) override {
+      auto result = nvenc->encode_frame(frame_index, force_idr);
+      force_idr = false;
+      return result;
+    }
+
+    int
+    convert(platf::img_t &img_base) override {
+      return base.convert(img_base);
```

Comment:
Nice refactoring. This will make it pretty simple if we want to have NVENC directly consume RGB frames or if we need to go that route with AMF for AMD cards.